qutebrowser/completion/models/sqlmodel.py

```diff
@@ -0,0 +1,273 @@
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
+
+"""A completion model backed by SQL tables."""
+
+import re
+
+from PyQt5.QtCore import Qt, QModelIndex, QAbstractItemModel
+from PyQt5.QtSql import QSqlQueryModel
+
+from qutebrowser.utils import log
+from qutebrowser.misc import sql
+
+
+class _SqlCompletionCategory(QSqlQueryModel):
+    def __init__(self, name, sort_by, sort_order, limit, select, where,
+                 columns_to_filter, parent=None):
+        super().__init__(parent=parent)
+        self.tablename = name
+
+        query = sql.run_query('select * from {} limit 1'.format(name))
+        self._fields = [query.record().fieldName(i) for i in columns_to_filter]
+
+        querystr = 'select {} from {} where ('.format(select, name)
+        querystr += ' or '.join("{} like ? escape '\\'".format(f)
+                                for f in self._fields)
+        querystr += ')'
+        if where:
+            querystr += ' and ' + where
+
+        if sort_by:
+            sortstr = 'asc' if sort_order == Qt.AscendingOrder else 'desc'
+            querystr += ' order by {} {}'.format(sort_by, sortstr)
+
+        if limit:
+            querystr += ' limit {}'.format(limit)
+
+        self._querystr = querystr
+        self.set_pattern('%')
+
+    def set_pattern(self, pattern):
+        query = sql.run_query(self._querystr, [pattern for _ in self._fields])
+        self.setQuery(query)
+
+
+class SqlCompletionModel(QAbstractItemModel):
+
+    """A sqlite-based model that provides data for the CompletionView.
+
+    This model is a wrapper around one or more sql tables. The tables are all
+    stored in a single database in qutebrowser's cache directory.
+
+    Top level indices represent categories, each of which is backed by a single
+    table. Child indices represent rows of those tables.
+
+    Class Attributes:
+        COLUMN_WIDTHS: The width percentages of the columns used in the
+                       completion view.
+
+    Attributes:
+        column_widths: The width percentages of the columns used in the
+                        completion view.
+        columns_to_filter: A list of indices of columns to apply the filter to.
+        pattern: Current filter pattern, used for highlighting.
+        _categories: The category tables.
+    """
+
+    def __init__(self, column_widths=(30, 70, 0), columns_to_filter=None,
+                 delete_cur_item=None, parent=None):
+        super().__init__(parent)
+        self.columns_to_filter = columns_to_filter or [0]
+        self.column_widths = column_widths
+        self._categories = []
+        self.srcmodel = self  # TODO: dummy for compat with old API
```

Comment:
The problem is that sometimes our model is wrapped and sometimes it isn't.
I think I could with the 'hybrid' completion model I talked about using for searchengine completion. It would look like:

    model = new CompletionModel()
    model.add_table('History')
    model.add_table('Quickmarks')
    model.add_table('Bookmarks')
    # this would add an item model wrapped in a sortfilter
    model.add_list('Searchengines', [('ddg', '...'), ...])

Then `set_pattern` would pass through to each of the wrapped models, not caring whether they're sql-based or not.

Then I could also eliminate [this awkward code](https://github.com/qutebrowser/qutebrowser/pull/2295/commits/d2943f419afebf37a5e1a1be21cf1dafe1c96aca#diff-e16711247709fa99263c3b02f61faa2b).